Add `isSoftDeleted` helper method to model instances

jsatk opened the pull request:
### Pull Request check-list

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)? **I need to fix failing tests**
- [x] Does your issue contain a link to existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Have you added an entry under `Future` in the changelog?

### Description of change

Since users can rename the `deletedAt` field when creating a model it is helpful to have a `isSoftDeleted` so other users of the DB don't have to hard code the custom `deletedAt` field name throughout their codebase.

Addresses: https://github.com/sequelize/sequelize/issues/7408

jsatk commented:
The test suite seems non-deterministic. It was passing. Adore and I pushed copy-only changes and now it's failing. I'd love to help fix if possible but a bit stumped st the moment.

janmeier commented on test/integration/instance.test.js:
```diff
@@ -2076,6 +2076,69 @@ describe(Support.getTestDialectTeaser('Instance'), () => {
+        username: { type: DataTypes.STRING }
+      }, { paranoid: true });
+
+      this.ParanoidUser.hasOne(this.ParanoidUser);
```
Is this relation needed?

jsatk replied:
I'm not entirely sure but I believe so. https://github.com/sequelize/sequelize/pull/7531/files/354210b1992afccaae8fdb270fddcb8e4752923f#diff-2756a0ac12fc81f7843fd569393b5966R1655 has the same code and I found that I needed that logic here.

jsatk replied:
I've removed it and all tests continue to pass. It seems it is not needed.

janmeier commented on test/unit/instance/is-soft-deleted.test.js:
```diff
@@ -0,0 +1,60 @@
+    it('should throw if paranoid is set to false', function() {
+      expect(() => {
+        this.user.isSoftDeleted();
+      }).to.throw();
```
`to.throw('Model is not paranoid')`

jsatk replied:
Done! Will amend fix to this commit.

jsatk commented:
@janmeier thank you! I responded to one and fixed the other. Thanks for the review. Excited to contribute. 😄

felixfbecker commented on test/integration/instance.test.js:
```diff
@@ -2076,6 +2076,69 @@ describe(Support.getTestDialectTeaser('Instance'), () => {
+    });
+
+    it('returns false if user is not soft deleted', function() {
+      const self = this;
```
Don't use `self`

jsatk replied:
Good call. It's not needed due to arrow function.

felixfbecker commented on lib/model.js:
```diff
@@ -3747,6 +3747,24 @@ class Model {
+   *
+   * @returns {Boolean}
+   */
+  isSoftDeleted() {
```
Yeah, shouldn't the function just check if `deletedAt` is set and less than current datetime? The default value logic looks weird to me

jsatk replied:
@felixfbecker The logic is meant to get the value of the property being used for soft delete (which by default is `deletedAt`. 

@sushantdhiman I didn't consider future deletes. I will update the logic to return true if the value of the soft delete property is same as or before now.

jsatk commented on test/integration/instance.test.js:
```diff
@@ -2076,6 +2076,69 @@ describe(Support.getTestDialectTeaser('Instance'), () => {
     });
   });
 
+  describe('isSoftDeleted', () => {
```
According to the linter in sequelize this is incorrect. It asks that you use arrow functions when possible over `function` and only use `function` when scope of `this` is a concern. 

`Unexpected function expression. (prefer-arrow-callback)`.

See https://github.com/sequelize/sequelize/blob/master/.eslintrc.json#L15

jsatk commented on lib/model.js:
```diff
@@ -3747,6 +3747,24 @@ class Model {
+    const deletedAtAttribute = this.constructor.rawAttributes[this.constructor._timestampAttributes.deletedAt];
+    const notSoftDeleted = deletedAtAttribute.hasOwnProperty('defaultValue') ? deletedAtAttribute.defaultValue : null;
+
+    return this.getDataValue(this.constructor._timestampAttributes.deletedAt) !== notSoftDeleted;
```
@sushantdhiman Will do! What is the difference between `this.get` and `this.getDataValue`? I see both being used in the codebase.

jsatk commented:
@felixfbecker @sushantdhiman I've addressed all comments and added in a few more unit tests. Please let me know if there's anything else. Thanks for the review. :smile:

janmeier commented:
@jsatk Fails for sqlite and mssql

``` 
  1) [SQLITE] Instance isSoftDeleted returns true if user is soft deleted:
     AssertionError: expected false to be true
      at null.<anonymous> (node_modules/chai/lib/chai/core/assertions.js:278:10)
      at addProperty (node_modules/chai/lib/chai/utils/addProperty.js:43:29)
      at doAsserterAsyncAndAddThen (node_modules/chai-as-promised/lib/chai-as-promised.js:296:33)
      at null.<anonymous> (node_modules/chai-as-promised/lib/chai-as-promised.js:286:25)
      at Object.defineProperty.get (node_modules/chai/lib/chai/utils/overwriteProperty.js:50:37)
      at test/integration/instance.test.js:2100:51
  From previous event:
      at Promise.then (node_modules/cls-bluebird/lib/shimMethod.js:38:20)
  2) [SQLITE] Instance isSoftDeleted works with custom `deletedAt` field name:
     AssertionError: expected false to be true
      at null.<anonymous> (node_modules/chai/lib/chai/core/assertions.js:278:10)
      at addProperty (node_modules/chai/lib/chai/utils/addProperty.js:43:29)
      at doAsserterAsyncAndAddThen (node_modules/chai-as-promised/lib/chai-as-promised.js:296:33)
      at null.<anonymous> (node_modules/chai-as-promised/lib/chai-as-promised.js:286:25)
      at Object.defineProperty.get (node_modules/chai/lib/chai/utils/overwriteProperty.js:50:37)
      at test/integration/instance.test.js:2127:53
  From previous event:
      at Promise.then (node_modules/cls-bluebird/lib/shimMethod.js:38:20)
```

jsatk commented:
Apologies.

I saw that this failed the other day and haven't had time to fix.  Will hopefully fix sometime this week.

On 05/03, Jan Aagaard Meier wrote:
>@jsatk Fails for sqlite and mssql
>
>```
>  1) [SQLITE] Instance isSoftDeleted returns true if user is soft deleted:
>     AssertionError: expected false to be true
>      at null.<anonymous> (node_modules/chai/lib/chai/core/assertions.js:278:10)
>      at addProperty (node_modules/chai/lib/chai/utils/addProperty.js:43:29)
>      at doAsserterAsyncAndAddThen (node_modules/chai-as-promised/lib/chai-as-promised.js:296:33)
>      at null.<anonymous> (node_modules/chai-as-promised/lib/chai-as-promised.js:286:25)
>      at Object.defineProperty.get (node_modules/chai/lib/chai/utils/overwriteProperty.js:50:37)
>      at test/integration/instance.test.js:2100:51
>  From previous event:
>      at Promise.then (node_modules/cls-bluebird/lib/shimMethod.js:38:20)
>  2) [SQLITE] Instance isSoftDeleted works with custom `deletedAt` field name:
>     AssertionError: expected false to be true
>      at null.<anonymous> (node_modules/chai/lib/chai/core/assertions.js:278:10)
>      at addProperty (node_modules/chai/lib/chai/utils/addProperty.js:43:29)
>      at doAsserterAsyncAndAddThen (node_modules/chai-as-promised/lib/chai-as-promised.js:296:33)
>      at null.<anonymous> (node_modules/chai-as-promised/lib/chai-as-promised.js:286:25)
>      at Object.defineProperty.get (node_modules/chai/lib/chai/utils/overwriteProperty.js:50:37)
>      at test/integration/instance.test.js:2127:53
>  From previous event:
>      at Promise.then (node_modules/cls-bluebird/lib/shimMethod.js:38:20)
>```
>
>--
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/sequelize/sequelize/pull/7531#issuecomment-298841883

jsatk commented:
@janmeier Any idea why the test suite date time is set to `1970-01-01T00:00:00.000Z`? Both my `deleted_at` and `now` values have a date time of `1970-01-01T00:00:00.000Z`. Any help? :)

felixfbecker commented on lib/model.js:
```diff
@@ -3747,6 +3747,38 @@ class Model {
+    const deletedAtAttribute = this.constructor.rawAttributes[this.constructor._timestampAttributes.deletedAt];
+    const defaultValue = deletedAtAttribute.hasOwnProperty('defaultValue') ? deletedAtAttribute.defaultValue : null;
+    const deletedAt = this.get(this.constructor._timestampAttributes.deletedAt);
+    const isNotSet = deletedAt === defaultValue;
```
Avoid booleans with negatives, it's hard to read something like `!isNotSet`. Prefer `isSet`

jsatk replied:
Done!

jsatk commented:
@felixfbecker @janmeier @sushantdhiman I believe this PR is good to go. Any other concerns?
